### PR TITLE
Use gin router everywhere

### LINKED ISSUE #439: [Server] integrate open telemetry using gin
https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/github.com/gin-gonic/gin/otelgin/example/server.go

### PR DESCRIPTION
**Pull request by decentralgabe**
fixes #439, #440
- refactor our service routers and how we return errs/responses to use gin
- renamed issuing to issuance to be in line with the rest of the apis
- updated all tests

### CODE REVIEW

File: pkg/server/router/did.go [hunk 1]
```diff
@@ -173,7 +178,7 @@ func optionsToType(options any, out any) error {
 }
 
 type GetDIDByMethodResponse struct {
-	DID didsdk.Document `json:"did,omitempty"`
+	DID didsdk.Document `json:"did"`
```
Comment (decentralgabe) [inline pkg/server/router/did.go:181]: making this consistent with other apis

File: pkg/server/server.go [hunk 2]
```diff
@@ -52,247 +50,247 @@ type SSIServer struct {
 }
 
 // NewSSIServer does two things: instantiates all service and registers their HTTP bindings
-func NewSSIServer(shutdown chan os.Signal, config config.SSIServiceConfig) (*SSIServer, error) {
+func NewSSIServer(shutdown chan os.Signal, cfg config.SSIServiceConfig) (*SSIServer, error) {
```
Comment (decentralgabe) [inline pkg/server/server.go:53]: main set of changes here

**Review by andresuribe87 — COMMENTED** [review]
This PR is lumping way too many changes in it. There are too many things going on that aren't related to routing. Can you separate into smaller ones?

**Comment by decentralgabe** [comment]
@andresuribe87 I'll remove renaming issuance the rest are related

File: cmd/ssiservice/main.go [hunk 3]
```diff
@@ -189,24 +193,17 @@ func configureLogger(level, location string) *os.File {
...
+	logrus.SetOutput(os.Stdout)
 	if location != "" {
 		logFile := location + "/" + config.ServiceName + "-" + time.Now().Format(time.RFC3339) + ".log"
 		file, err := os.OpenFile(logFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
 		if err != nil {
 			logrus.WithError(err).Warn("failed to create logs file, using default stdout")
 		} else {
-			logrus.SetOutput(file)
+			mw := io.MultiWriter(os.Stdout, file)
+			logrus.SetOutput(mw)
```
Comment (decentralgabe) [inline cmd/ssiservice/main.go:206]: this is the fix for #462

File: pkg/server/middleware/errors.go [hunk 4]
```diff
@@ -13,31 +16,26 @@ import (
...
-		}
-
 		errors := c.Errors.ByType(gin.ErrorTypeAny)
 		if len(errors) > 0 {
 			// check if there's a shutdown-worthy error
 			for _, e := range errors {
 				if framework.IsShutdown(e.Err) {
-					c.Set(framework.ShutdownErrorState.String(), e.Err)
+					logrus.Errorf("%s : SHUTDOWN ERROR : %v", span.SpanContext().TraceID().String(), e)
+					shutdown <- syscall.SIGTERM
```
Comment (andresuribe87) [inline pkg/server/middleware/errors.go:32]: What's the use case to allow handlers to shut down the server?

File: pkg/server/server_test.go [hunk 5]
```diff
@@ -343,3 +334,7 @@ func testWebhookRouter(t *testing.T, bolt storage.ServiceStorage) *router.Webhoo
 
 	return webhookRouter
 }
+
+func is2xxResponse(statusCode int) bool {
+	return statusCode/100 == 2
+}
```
Comment (andresuribe87) [inline pkg/server/server_test.go:340]: We have this defined in multiple places, perhaps it's time to drop this into util?

File: pkg/service/credential/service.go [hunk 6]
```diff
@@ -193,7 +193,7 @@ func (s Service) createCredentialBusinessLogic(ctx context.Context, request Crea
 	}
 
 	if err := builder.SetIssuanceDate(time.Now().Format(time.RFC3339)); err != nil {
-		errMsg := fmt.Sprintf("could not set credential issuance date")
+		errMsg := fmt.Sprintf("could not set credential issuing date")
```
Comment (andresuribe87) [inline pkg/service/credential/service.go:196]: Was this intentional?

File: pkg/service/issuance/model.go [hunk 7]
```diff
@@ -108,6 +108,6 @@ func (r ListIssuanceTemplatesRequest) Validate() error {
 }
 
 type ListIssuanceTemplatesResponse struct {
-	// The issuance templates that satisfy the query conditions.
+	// The issuing templates that satisfy the query conditions.
```
Comment (andresuribe87) [inline pkg/service/issuance/model.go:111]: Intentional?

File: pkg/service/issuance/storage.go
Comment (andresuribe87) [inline pkg/service/issuance/storage.go:1]: I think this file didn't need to change?

File: sip/sips/sip5/README.md
Comment (andresuribe87) [inline sip/sips/sip5/README.md:1]: I think these might not be intentional.

File: pkg/server/framework/errors.go [hunk 8]
```diff
@@ -11,8 +15,8 @@ type FieldError struct {
 // ErrorResponse is the structure of response error payloads sent back to the requester
 // when validation of a request payload fails.
 type ErrorResponse struct {
-	Error  string       `json:"error"`
-	Fields []FieldError `json:"fields,omitempty"`
+	Error  string `json:"error"`
+	Fields string `json:"fields,omitempty"`
```
Comment (andresuribe87) [inline pkg/server/framework/errors.go:19]: Curious why this change?

File: config/dev.toml [hunk 9]
```diff
@@ -34,29 +33,16 @@ storage = "bolt"
...
 password = "default-password"
-# master_key_uri = "gcp-kms://projects/*/locations/*/keyRings/*/cryptoKeys/*"
-# kms_credentials_path = "credentials.json"
 
 [services.did]
 name = "did"
 methods = ["key", "web"]
 local_resolution_methods = ["key", "web", "pkh", "peer"]
-universal_resolver_url = "http://localhost:8088"
+universal_resolver_url = "https://dev.uniresolver.io/"
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply config/dev.toml:45]: if you're running locally (just a single go process) you won't have a uni resolver running - that's `dev.toml`
`test.toml` is used when you're running docker-compose so you will have the local uni resolver

File: config/test.toml [hunk 10]
```diff
@@ -18,7 +19,7 @@ shutdown_timeout = 5000000000
 
 log_location = "log"
 # options: trace, debug, info, warn, error, fatal, panic
-log_level = "warn"
+log_level = "debug"
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply config/test.toml:22]: yeah I'll change this back

File: integration/credential_manifest_integration_test.go [hunk 11]
```diff
@@ -32,20 +32,6 @@ func TestCreateIssuerDIDKeyIntegration(t *testing.T) {
 	SetValue(credentialManifestContext, "issuerKID", issuerKID)
 }
 
-func TestResolveIONDIDIntegration(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply integration/credential_manifest_integration_test.go:35]: forgot to move this to the ion integration test, will add it back there

File: pkg/server/middleware/errors.go [hunk 12]
```diff
@@ -13,31 +16,26 @@ import (
...
-			return
-		}
-
 		errors := c.Errors.ByType(gin.ErrorTypeAny)
 		if len(errors) > 0 {
 			// check if there's a shutdown-worthy error
 			for _, e := range errors {
 				if framework.IsShutdown(e.Err) {
-					c.Set(framework.ShutdownErrorState.String(), e.Err)
+					logrus.Errorf("%s : SHUTDOWN ERROR : %v", span.SpanContext().TraceID().String(), e)
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply pkg/server/middleware/errors.go:31]: yes

[the remark this answers is not in the document]
Comment (decentralgabe) [reply pkg/server/middleware/errors.go:31]: yes

File: pkg/server/middleware/webhook.go [hunk 13]
```diff
@@ -24,9 +24,11 @@ func (rw *responseWriter) Write(b []byte) (int, error) {
 	return rw.ResponseWriter.Write(b)
 }
 
+type Publisher func(c *gin.Context, noun webhook.Noun, verb webhook.Verb, reader io.Reader)
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply pkg/server/middleware/webhook.go:27]: ah this was a piece of refactoring I've abandoned, will remove

File: pkg/server/middleware/webhook.go [hunk 14]
```diff
@@ -42,7 +44,6 @@ func Webhook(webhookService svcframework.Service, noun webhook.Noun, verb webhoo
 		}
 
 		// publish the webhook
-		whService := webhookService.(*webhook.Service)
-		go whService.PublishWebhook(noun, verb, buf)
+		go webhookService.PublishWebhook(c, noun, verb, buf)
```
Comment (nitro-neal) [inline pkg/server/middleware/webhook.go:47]: be careful passing the context here, I purposly removed it because on doing an integration test of webhooks, the context closes in the middle of this go routine, so it kills the POST

File: pkg/server/server_test.go (hunk 5 above)
Comment (decentralgabe) [reply pkg/server/server_test.go:340]: done

File: pkg/service/credential/service.go (hunk 6 above)
Comment (decentralgabe) [reply pkg/service/credential/service.go:196]: will revert

File: pkg/service/did/ion_test.go [hunk 15]
```diff
@@ -74,6 +74,11 @@ func TestIONHandler(t *testing.T) {
 	})
 
 	t.Run("Test Create DID", func(tt *testing.T) {
+		gock.New("https://ion.tbddev.org").
+			Post("/operations").
+			Reply(200)
+		defer gock.Off()
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply pkg/service/did/ion_test.go:80]: otherwise the `handler.CreateDID` method below will make a call to the real http endpoint - and this is not an integration test

File: pkg/service/issuance/model.go (hunk 7 above)
Comment (decentralgabe) [reply pkg/service/issuance/model.go:111]: reverted

File: sip/sips/sip5/README.md
[the remark this answers is not in the document]
Comment (decentralgabe) [reply sip/sips/sip5/README.md:1]: reverted

File: pkg/service/manifest/service.go [hunk 16]
```diff
@@ -330,7 +330,7 @@ func (s Service) ProcessApplicationSubmission(ctx context.Context, request model
 	return operation.ServiceModel(*storedOp)
 }
 
-// attemptAutomaticIssuance checks if there is an issuance template for the manifest, and if so,
+// attemptAutomaticIssuance checks if there is an issuing template for the manifest, and if so,
```
Comment (nitro-neal) [inline pkg/service/manifest/service.go:333]: Not sure if you want it issuance or issuing in this file

File: pkg/server/server.go [hunk 17]
```diff
@@ -52,247 +50,248 @@ type SSIServer struct {
...
-	s.Handle(http.MethodGet, handlerPath, didRouter.GetDIDMethods)
-	s.Handle(http.MethodPut, path.Join(handlerPath, "/:method"), didRouter.CreateDIDByMethod, middleware.Webhook(webhookService, webhook.DID, webhook.Create))
-	s.Handle(http.MethodGet, path.Join(handlerPath, "/:method"), didRouter.GetDIDsByMethod)
-	s.Handle(http.MethodGet, path.Join(handlerPath, "/:method/:id"), didRouter.GetDIDByMethod)
-	s.Handle(http.MethodDelete, path.Join(handlerPath, "/:method/:id"), didRouter.SoftDeleteDIDByMethod)
-
-	s.Handle(http.MethodGet, path.Join(path.Join(handlerPath, ResolverPrefix), "/:id"), didRouter.ResolveDID)
+	didAPI := rg.Group(DIDsPrefix)
+	didAPI.GET("", didRouter.GetDIDMethods)
+	didAPI.PUT("/:method", didRouter.CreateDIDByMethod, middleware.Webhook(webhookService, webhook.DID, webhook.Create))
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply pkg/server/server.go:142]: gin will order appropriately depending on whether you call `c.Next` in the middleware; I'll update regardless to be clear

File: pkg/service/webhook/service.go [hunk 18]
```diff
@@ -173,13 +175,14 @@ func (s Service) GetSupportedVerbs() GetSupportedVerbsResponse {
 	return GetSupportedVerbsResponse{Verbs: []Verb{Create, Delete}}
 }
 
-func (s Service) PublishWebhook(noun Noun, verb Verb, payloadReader io.Reader) {
-	ctx, cancel := context.WithTimeout(context.Background(), s.timeoutDuration)
+// TODO: consider returning an error to be handled by the gin middleware
```
Comment (nitro-neal) [inline pkg/service/webhook/service.go:178]: Be careful I tried thee same thing here, the c Context finishes during the middle of the PublishWebhook call and it never does a POST at the bottom, can do an integration test by creaing a nodejs server to recieve webhook, or can pull down this branch of playground to test webhook - https://github.com/TBD54566975/ssi-playground/tree/webhook-demo

File: pkg/server/middleware/webhook.go (hunk 14 above)
Comment (decentralgabe) [reply pkg/server/middleware/webhook.go:44]: we need it now to propagate any errors that are encountered, it's the gin context so may be different behavior. is there a way I can test the cancelling behavior?

**Review by nitro-neal — APPROVED** [review]
Just double check the webhook thing, awesome job! Happy to help with testing if needed

File: pkg/service/manifest/service.go (hunk 16 above)
Comment (decentralgabe) [reply pkg/service/manifest/service.go]: thanks updated

File: pkg/server/middleware/webhook.go (hunk 14 above)
Comment (andresuribe87) [reply pkg/server/middleware/webhook.go:44]: I agree that you shouldn't be passing it. The gin context shouldn't apply to the webhook publishing.

File: pkg/service/webhook/service.go (hunk 18 above)
Comment (andresuribe87) [reply pkg/service/webhook/service.go:178]: It doesn't cover it because this is a cancellation issue. You'd need to time things perfectly. 

The test could be modified to make it happen, but I don't this it's worth it.

**Review by andresuribe87 — APPROVED** [review]
Looks good sans passing the same context to the webhook. You probably want to call `c.Copy`
